src/bot/snoopiBot.jl

```diff
@@ -0,0 +1,109 @@
+
+################################################################
+"""
+    @snoopiBot config::BotConfig snoopScript
+
+macro that generates precompile files and includes them in the package. Calls other bot functions.
+
+# Examples
+
+`@snoopiBot` the examples that call the package functions.
+
+```julia
+using SnoopCompile
+
+@snoopiBot "MatLang" begin
+  using MatLang
+  examplePath = joinpath(dirname(dirname(pathof(MatLang))), "examples")
+  include(joinpath(examplePath,"Language_Fundamentals", "usage_Entering_Commands.jl"))
+  include(joinpath(examplePath,"Language_Fundamentals", "usage_Matrices_and_Arrays.jl"))
+  include(joinpath(examplePath,"Language_Fundamentals", "Data_Types", "usage_Numeric_Types.jl"))
+end
+```
+"""
+macro snoopiBot(config::BotConfig, snoopScript)
+
+    packageName = config.packageName
+    blacklist = config.blacklist
+    subst = config.subst
+    ################################################################
+    packagePath = joinpath(pwd(),"src","$packageName.jl")
+    precompilePath, precompileFolder = precompilePather(packageName)
+
+    quote
+        packageSym = Symbol($packageName)
+        ################################################################
+        using SnoopCompile
+        ################################################################
+        precompileDeactivator($packagePath, $precompilePath);
+        ################################################################
+
+        ### Log the compiles
+        data = @snoopi begin
+            $(esc(snoopScript))
+        end
+
+        ################################################################
+        ### Parse the compiles and generate precompilation scripts
+        pc = SnoopCompile.parcel(data, subst = $subst, blacklist = $blacklist)
+        onlypackage = Dict( packageSym => sort(pc[packageSym]) )
+        SnoopCompile.write($precompileFolder,onlypackage)
+        ################################################################
+        precompileActivator($packagePath, $precompilePath)
+    end
+
+end
+
+macro snoopiBot(packageName::String, snoopScript)
+    config = BotConfig(packageName)
+    return quote
+        @snoopiBot $config $(esc(snoopScript))
+    end
+end
+macro snoopiBot(configExpr, snoopScript)
+    config = eval(configExpr)
```

Comment:
What's this about? Can't we just pass back the expression and have it eval normally?

Reply:
I use this line to get `BotConfig` type and dispatch on it for the macro. Because macros can't detect types, I have to eval it.